dynamic_array.c

```diff
@@ -143,20 +150,22 @@ void buffer_push_back(struct buffer *inst, void *item)
             --inst->length;
```

Comment:
Currently, you increase the length (in addition to having the previous length stored in a variable), then un-do the increase if item is NULL or if we need to grow the buffer.  It's best to avoid "undo" code as it can easily lead to bugs during refactoring and future maintenance.  Instead, only alter `inst->length` if we push an item, and if `item == NULL` then it might be better to zero out a new item at the end of the buffer rather than silently failing [note that I don't think I had specified how we should handle this case in the original specification].